File: vs2003_cd02/Program Files/Microsoft Visual Studio .NET 2003/SDK/v1.1/QuickStart/winforms/samples/writingcontrols/simplecontrol/cs/control/simplecontrol.cs
//-----------------------------------------------------------------------
//  This file is part of the Microsoft .NET SDK Code Samples.
// 
//  Copyright (C) Microsoft Corporation.  All rights reserved.
// 
//This source code is intended only as a supplement to Microsoft
//Development Tools and/or on-line documentation.  See these other
//materials for detailed information regarding Microsoft code samples.
// 
//THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
//KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//PARTICULAR PURPOSE.
//-----------------------------------------------------------------------


namespace Microsoft.Samples.WinForms.Cs.SimpleControl {
    using System;
    using System.ComponentModel;
    using System.Windows.Forms;
    using System.Drawing;

    [
    DefaultProperty("DrawingMode"),
    DefaultEvent("DrawingModeChanged"),
    ]
    public class SimpleControl : System.Windows.Forms.Control {

        private DrawingMode drawingMode ;
        private System.EventHandler onDrawingModeChanged;

        //*** Constructors

        public SimpleControl() :base() {

            //Initialise drawingMode
            drawingMode = DrawingMode.Happy;

            //Initialise BackColor and ForeColor based on DrawingMode
            SetColors();

            //Make sure the control repaints as it is resized
            SetStyle(ControlStyles.ResizeRedraw, true);
        }


        //*** Properties

        //Remove the BackColor property from the properties window
        [Browsable(false)]
        public override Color BackColor {
            get {return base.BackColor;}
            set {}
        }


        //DrawingMode - controls how the control paints
        [
            Category("Appearance"),
            Description("Controls how the control paints"),
            DefaultValue(DrawingMode.Happy),
            Bindable(true),
        ]
        public DrawingMode DrawingMode {
            get { return drawingMode;}
            set {
                drawingMode=value;

                //Raise property changed event for DrawingMode
                OnDrawingModeChanged(EventArgs.Empty);
            }
        }


        //Remove the ForeColor property from the properties window
        [Browsable(false)]
        public override Color ForeColor  {
            get {return base.ForeColor;}
            set {}
        }


        //*** Events

        //Handle the paint event
        protected override void OnPaint(PaintEventArgs e) {

            e.Graphics.FillRectangle(new SolidBrush(BackColor), ClientRectangle);

            Size textSize = e.Graphics.MeasureString(Text, Font).ToSize();

            float xPos = (ClientRectangle.Width/2) - (textSize.Width/2);
            float yPos = (ClientRectangle.Height/2) - (textSize.Height/2);

            e.Graphics.DrawString(Text,
                                  Font,
                                  new SolidBrush(ForeColor),
                                  xPos, yPos);

        }

        protected override void OnTextChanged(EventArgs e) {
            base.OnTextChanged(e);
            Invalidate();
        }

        //DrawingModeChanged Event
        [Description("Raised when the DrawingMode changes")]
        public event EventHandler DrawingModeChanged {
            add {
                onDrawingModeChanged += value;
            }
            remove {
                onDrawingModeChanged -= value;
            }
        }

        protected virtual void OnDrawingModeChanged(EventArgs e) {

            //Set BackColor and ForeColor based on DrawingMode
            SetColors();

            Invalidate();

            if (onDrawingModeChanged != null) onDrawingModeChanged.Invoke(this, e);
        }


        //Set the ForeColor and BackColor based on the value of DrawingMode
        private void SetColors() {

            switch (drawingMode) {

               case DrawingMode.Happy:
                   base.BackColor = Color.Yellow ;
                   base.ForeColor = Color.Green ;
                   break ;

               case DrawingMode.Sad:
                   base.BackColor = Color.LightSlateGray ;
                   base.ForeColor = Color.White ;
                   break ;

              case DrawingMode.Angry:
                  base.BackColor = Color.Red ;
                  base.ForeColor = Color.Teal ;
                  break ;

               default:
                   base.BackColor = Color.Black ;
                   base.ForeColor = Color.White ;
					break;
           }

        }


    }
}

File: vs2003_cd02/Program Files/Microsoft Visual Studio .NET 2003/SDK/v1.1/QuickStart/winforms/samples/writingcontrols/simplecontrol/vb/control/SimpleControl.vb
'-----------------------------------------------------------------------
'  This file is part of the Microsoft .NET SDK Code Samples.
' 
'  Copyright (C) Microsoft Corporation.  All rights reserved.
' 
'This source code is intended only as a supplement to Microsoft
'Development Tools and/or on-line documentation.  See these other
'materials for detailed information regarding Microsoft code samples.
' 
'THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
'KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'PARTICULAR PURPOSE.
'-----------------------------------------------------------------------


Imports System
Imports System.ComponentModel
Imports System.Windows.Forms
Imports System.Drawing

Namespace Microsoft.Samples.WinForms.VB.SimpleControl

    <DefaultProperty("DrawingMode"), DefaultEvent("DrawingModeChanged")> _
        Public Class SimpleControl

        Inherits System.Windows.Forms.Control

        Private Sub InitializeComponent()

        End Sub

        Private myDrawingMode As DrawingModeStyle
        Private myOnDrawingModeChanged As EventHandler

        '*** Constructors

        Public Sub New()
            MyBase.New()

            'Initialise drawingMode
            myDrawingMode = DrawingModeStyle.Happy

            'Initialise BackColor and ForeColor based on DrawingMode
            SetColors()

            'Make sure the control repaints as it is resized
            SetStyle(ControlStyles.ResizeRedraw, True)

        End Sub


        '*** Properties

        'Remove the BackColor property from the properties window
        <Browsable(False)> Public Overrides Property BackColor() As Color
            Get
                Return MyBase.BackColor
            End Get

            Set(ByVal Value As Color)
                'No Action
            End Set

        End Property


        'DrawingMode - controls how the control paints
        <Category("Appearance"), _
         Description("Controls how the control paints"), _
         DefaultValue(DrawingModeStyle.Happy), _
         Bindable(True)> _
        Public Property _
        DrawingMode() As DrawingModeStyle

            Get
                Return myDrawingMode
            End Get

            Set(ByVal Value As DrawingModeStyle)
                myDrawingMode = Value

                'Set BackColor and ForeColor based on DrawingMode
                SetColors()

                'Raise property changed event for DrawingMode
                OnDrawingModeChanged(EventArgs.Empty)

            End Set

        End Property


        'Remove the ForeColor property from the properties window
        <Browsable(False)> Public Overrides Property ForeColor() As Color
            Get
                Return MyBase.ForeColor
            End Get

            Set(ByVal Value As Color)
                'No Action
            End Set

        End Property



        '*** Events

        'Handle the paint event
        Protected Overrides Sub OnPaint(ByVal e As PaintEventArgs)

            e.Graphics.FillRectangle(New SolidBrush(BackColor), ClientRectangle)

            Dim textSize As SizeF = e.Graphics.MeasureString(Me.Text, Font)

            Dim xPos As Single = CSng((ClientRectangle.Width / 2) - (textSize.Width / 2))
            Dim yPos As Single = CSng((ClientRectangle.Height / 2) - (textSize.Height / 2))

            e.Graphics.DrawString(Me.Text, Font, New SolidBrush(ForeColor), xPos, yPos)

        End Sub

        Protected Overrides Sub OnTextChanged(ByVal e As EventArgs)
            MyBase.OnTextChanged(e)
            Invalidate()
        End Sub


        'Catch property changed event for DrawingMode to fire DrawingMoe changed
        'and repaint the control
        Protected Overridable Sub OnDrawingModeChanged(ByVal E As EventArgs)
            'Set BackColor and ForeColor based on DrawingMode
            SetColors()
            Invalidate()
            If Not (myOnDrawingModeChanged Is Nothing) Then myOnDrawingModeChanged.Invoke(Me, E)
        End Sub

        'DrawingModeChanged Event
            <Description("Raised when the DrawingMode changes")> _
            Public Event _
            DrawingModeChanged(ByVal sender As Object, ByVal ev As EventArgs) 'As EventHandler

        'Set the ForeColor and BackColor based on the value of DrawingMode
        Private Sub SetColors()

            Select Case myDrawingMode

                Case DrawingModeStyle.Happy
                    MyBase.BackColor = Color.Yellow
                    MyBase.ForeColor = Color.Green

                Case DrawingModeStyle.Sad
                    MyBase.BackColor = Color.LightSlateGray
                    MyBase.ForeColor = Color.White

                Case DrawingModeStyle.Angry
                    MyBase.BackColor = Color.Red
                    MyBase.ForeColor = Color.Teal

                Case Else
                    MyBase.BackColor = Color.Black
                    MyBase.ForeColor = Color.White

            End Select

        End Sub

    End Class

End Namespace
